### PR TITLE
handle delayed AAD resources in deployments

### PR DESCRIPTION
**Pull request by bmc-msft**
This handles exceptions generated during deployment when newly created AAD resources are not yet available for use by the ARM deployment.

An example error message:

```
Message: {'error': {'code': 'PrincipalNotFound', 'message': 'Principal HEXGUID does not exist in the directory GUID.'}}
```

### CODE REVIEW

File: src/deployment/deploy.py [hunk 1]
```diff
@@ -384,16 +384,36 @@ def deploy_template(self) -> None:
...
+            try:
+                result = client.deployments.create_or_update(
+                    self.resource_group, gen_guid(), deployment
+                ).result()
+                if result.properties.provisioning_state != "Succeeded":
+                    logger.error(
+                        "error deploying: %s",
+                        json.dumps(result.as_dict(), indent=4, sort_keys=True),
+                    )
+                    sys.exit(1)
```
Comment (ranweiler) [inline src/deployment/deploy.py:402]: Is there a reason we `exit(1)` here, but throw an exception in other error paths? Is the discrepancy just a refactoring artifact?

File: src/deployment/registration.py [hunk 2]
```diff
@@ -235,7 +235,10 @@ def add_application_password(app_object_id: UUID) -> Tuple[str, str]:
             return add_application_password_impl(app_object_id)
         except GraphQueryError as err:
             error = err
-            logging.warning("unable to create app password: %s", err.message)
+            if "Request_ResourceNotFound" in repr(err):
```
Comment (ranweiler) [inline src/deployment/registration.py:238]: Can we cite something for the origin of this search substring? Or at least comment on where it is empirically derived?

Same applies to `PrincipalNotFound`, above.

**Review by ranweiler — APPROVED** [review]

File: src/deployment/deploy.py (hunk 1 above)
Comment (bmc-msft) [reply src/deployment/deploy.py:402]: This is a refactoring artifact.

Comment (bmc-msft) [reply src/deployment/deploy.py:402]: This is a refactoring artifact.

**Review by chkeita — APPROVED** [review]

File: src/deployment/registration.py (hunk 2 above)
Comment (bmc-msft) [reply src/deployment/registration.py:243]: I added some refs to how this is similar to az-cli

**Pull request merged**